src/librustdoc/html/format.rs

```diff
@@ -456,15 +456,26 @@ impl clean::GenericArgs {
     }
 }
 
-crate fn href(did: DefId, cx: &Context<'_>) -> Option<(String, ItemType, Vec<String>)> {
+// Possible errors when computing href link source for a `DefId`
+crate enum HrefError {
+    // `DefId` is in an unknown location. This seems to happen when building without dependencies
+    // but a trait from a dependency is still visible
+    UnknownLocation,
+    // Unavailable because private
+    Unavailable,
+    // Not in external cache, href link should be in same page
+    NotInExternalCache,
+}
+
+crate fn href(did: DefId, cx: &Context<'_>) -> Result<(String, ItemType, Vec<String>), HrefError> {
     let cache = &cx.cache();
     let relative_to = &cx.current;
     fn to_module_fqp(shortty: ItemType, fqp: &[String]) -> &[String] {
         if shortty == ItemType::Module { &fqp[..] } else { &fqp[..fqp.len() - 1] }
     }
 
     if !did.is_local() && !cache.access_levels.is_public(did) && !cache.document_private {
-        return None;
+        return Err(HrefError::Unavailable);
     }
 
     let (fqp, shortty, mut url_parts) = match cache.paths.get(&did) {
```

Comment:
Oof, having this be different from `exact_paths` is very concerning. There's a lot of cleanup rustdoc needs to do here :sweat_smile: not related to your PR though

Reply:
Opened https://github.com/rust-lang/rust/issues/86798 which is not exactly the same issue but sort of cloes.